apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4SamhsaMatcher.java

```diff
@@ -595,8 +595,12 @@ private boolean isSamhsaCptCode(CodeableConcept procedureConcept) {
      * Note: CPT codes represent a subset of possible HCPCS codes (but are the only
      * subset that we blacklist from).
      */
+    Set<String> codingSystems =
+        procedureConcept.getCoding().stream()
+            .map(coding -> coding.getSystem())
+            .collect(Collectors.toSet());
     for (Coding procedureCoding : procedureConcept.getCoding()) {
-      if (!TransformerConstants.CODING_SYSTEM_HCPCS.equals(procedureCoding.getSystem())) {
+      if (!codingSystems.contains(TransformerConstants.CODING_SYSTEM_HCPCS)) {
         throw new IllegalArgumentException();
       }
```

Comment:
This can be moved outside the for loop.

Reply:
Could this ever happen?  For `isSamhsaCptCode()` to get invoked, it would have had to pass the `hasHcpcsCode` check it would seem.

Reply:
Yeah at present this can't be triggered. For super safety you could make the argument that this method could be called from elsewhere in the future and to guard against improper use, but I've seen the argument made not to be overly defensive too.

Reply:
I'm also wondering if this should remain one method as well?  I like breaking things out into simple logical methods, but I'm wondering if this all falls under the same check concept?
```java
  protected boolean containsSamhsaProcedureCode(CodeableConcept procedureConcept) {
    Map<String, Coding> codings = procedureConcept.getCoding().stream()
            .collect(Collectors.toMap(Coding::getSystem, Function.identity()));

    if (codings.containsKey(TransformerConstants.CODING_SYSTEM_HCPCS)) {
      return isSamhsaCptCode(codings.get(TransformerConstants.CODING_SYSTEM_HCPCS)) // HCPCS Samhsa code
              // Or if there are extra unwanted systems
              || (!codings.keySet().equals(Set.of(TransformerConstants.CODING_SYSTEM_HCPCS))
                && !codings.keySet().equals(Set.of(TransformerConstants.CODING_SYSTEM_HCPCS, TransformerUtilsV2.calculateVariableReferenceUrl(CcwCodebookVariable.HCPCS_CD))));
    } else {
      return !codings.isEmpty();
    }
  }

Reply:
Just want to chime in on the general principle here: we should bias **heavily** towards being overly defensive in any/all of our systems. Extra assertions hurt **nothing** and often make the code much simpler for readers to reason about, which is a major win.

Reply:
As a single method, it wouldn't be an issue, I'm just wondering if that might serve better here, as the check is very specific to this one use case it seems like, and it's arguably not too over complicated as such.

Reply:
Resolved by making it one method :)